lib/Onyx.js

```diff
@@ -699,7 +699,10 @@ function initializeWithDefaultKeyStates() {
 
             const merged = lodashMerge(asObject, defaultKeyStates);
             cache.merge(merged);
-            _.each(merged, (val, key) => keyChanged(key, val));
+            _.each(merged, (val, key) => {
+                const currentValue = cache.getValue(key);
+                keyChanged(key, currentValue);
+            });
```

Comment:
**question:** It looks like `merged` is going to end up as the combined result of any existing values for these keys + the `defaultKeyStates` object? I might just be missing it, but why would it be necessary to ask the cache for this data if `merged` is the combined result already?

Reply:
In the case that I'm trying to fix, `merged` contains only the `defaultKeyStates` as the `Storage` is empty. However, the `cache` is not empty. So if the `defaultKeyStates` are not merged with the value in the `cache` then `keyChanged` is called without up to date values.

Reply:
>However, the cache is not empty.

**thought:** If we are signing out then both the cache and storage should be empty. Am I missing something?

Reply:
https://github.com/Expensify/react-native-onyx/blob/2d42566779884bb445aa59dd48b5bcdff6c0a4e6/lib/Onyx.js#L711-L721

Reply:
It is also surprising to me that the cache is not empty but the storage is. My guess is that `Storage.clear` takes a while to complete, and during that time the new user is logged in and lots of values are set in Onyx. 

When are values written from the cache to storage? Can the storage be out of date temporarily? I will try to look through the code and answer my own questions, but if you can give me a quick explanation / point me in the right direction that would be great.

Reply:
>My guess is that Storage.clear takes a while to complete, and during that time the new user is logged in and lots of values are set in Onyx.

That seems possible. Here's where we clear on log out:

https://github.com/Expensify/App/blob/f6dc28f6a6c3d110a65e5caa33f94d441be3b630/src/libs/actions/SignInRedirect.js#L22-L39

But in the case where we sign someone out and sign in a new person we might not be waiting for that `Onyx.clear()` before logic elsewhere signs in the user? Probably would check to see if this code:

https://github.com/Expensify/App/blob/f6dc28f6a6c3d110a65e5caa33f94d441be3b630/src/pages/LogInWithShortLivedTokenPage.js#L57-L62

Runs before this promise resolves:

https://github.com/Expensify/App/blob/f6dc28f6a6c3d110a65e5caa33f94d441be3b630/src/libs/actions/SignInRedirect.js#L28

Reply:
>When are values written from the cache to storage? 

I don't think values from the cache are even written to storage.

>Can the storage be out of date temporarily?

I'm not sure if I have enough context to explain a situation where the storage could be out of date. But it certainly seems possible.

Reply:
I have verified that signing in the new user occurs before Onyx.clear() finishes. Waiting to sign the user in until the storage clears could work, and I guess we could do that, but wouldn't it be better to not wait for that? I also think that this solution makes more sense. Please see the issue for more context and an example if you haven't seen that already.

Reply:
>Waiting to sign the user in until the storage clears could work, and I guess we could do that, but wouldn't it be better to not wait for that?

**clarification:** By better do we mean faster? Or something else. 

**thought:** Signing out a user and signing in a new one seems like a rare event. So maybe performance is not a concern (assuming that is the concern).

>I also think that this solution makes more sense.

I do see how it solves the immediate issue. But just want to make sure we're picking the best solution. I'm not sure if allowing data to be set while we are in the process of re-initializing storage makes sense. I think my expectation is that we'd defer those updates until the app has finished one "clean" cycle.

Mind if we go back to this question?

**Should we completely sign the previous user out before starting a new sign in process?**

Reply:
> clarification: By better do we mean faster? Or something else.

Yes. I think you're right that it probably wouldn't slow things down too much and that this flow is rare.

> I'm not sure if allowing data to be set while we are in the process of re-initializing storage makes sense.

It would be up to the user of Onyx to make sure that no data is set while re-initializing storage right? That seems slightly less robust. If any initialKey is set while re-initializing storage then it could send an out of date value to its subscribers. How do you make your whole app wait while Onyx clears and re-initializes?

> Should we completely sign the previous user out before starting a new sign in process?

Probably, yes. I will work on a PR to do that. If that works and we agree that it's a better solution, then I'm happy to do that instead.

Reply:
>It would be up to the user of Onyx to make sure that no data is set while re-initializing storage right?

Yep!

>That seems slightly less robust. If any initialKey is set while re-initializing storage then it could send an out of date value to its subscribers. How do you make your whole app wait while Onyx clears and re-initializes?

`Onyx.clear()` returns a promise so that would be my first instinct. But maybe there's a different solution. Might be something we can get some more opinions on too? I'm not 100% what the best answer would be, but just wanted to think about it for a second. Thanks for hearing me out!